homework6/oop_2.py

```diff
@@ -0,0 +1,199 @@
+
+HomeworkResult принимает объект автора задания, принимает исходное задание
+и его решение в виде строки
+Атрибуты:
+    homework - для объекта Homework, если передан не этот класс -  выкинуть
+    подходящие по смыслу исключение с сообщением:
+    "You gave a not Homework object"
+
+    solution - хранит решение ДЗ как строку
+    author - хранит объект Student
+    created - c точной датой и временем создания
+
+2. Если задание уже просрочено хотелось бы видеть исключение при do_homework,
+а не просто принт "You are late".
+Поднимайте исключение DeadlineError с сообщением "You are late" вместо print.
+
+3. Student и Teacher имеют одинаковые по смыслу атрибуты
+(last_name, first_name) - избавиться от дублирования с помощью наследования
+
+4.
+Teacher
+Атрибут:
+    homework_done - структура с интерфейсом как в словаря, сюда поподают все
+    HomeworkResult после успешного прохождения check_homework
+    (нужно гаранитровать остутствие повторяющихся результатов по каждому
+    заданию), группировать по экземплярам Homework.
+    Общий для всех учителей. Вариант ипользования смотри в блоке if __main__...
+Методы:
+    check_homework - принимает экземпляр HomeworkResult и возвращает True если
+    ответ студента больше 5 символов, так же при успешной проверке добавить в
+    homework_done.
+    Если меньше 5 символов - никуда не добавлять и вернуть False.
+
+    reset_results - если передать экземпряр Homework - удаляет только
+    результаты этого задания из homework_done, если ничего не передавать,
+    то полностью обнулит homework_done.
+
+PEP8 соблюдать строго.
+Всем перечисленным выше атрибутам и методам классов сохранить названия.
+К названием остальных переменных, классов и тд. подходить ответственно -
+давать логичные подходящие имена.
+"""
+import datetime
+from collections import defaultdict
+
+
+class Person:
+    """
+    Initiates attributes last_name and first_name
+    """
+
+    def __init__(self, first_name, last_name):
+        self.last_name = last_name
+        self.first_name = first_name
+
+
+class Homework:
+    """
+    Initiates:
+    Attributes text, deadline, created
+    Method is_active, that return True if deadline didn"t happen else return False
+    """
+
+    def __init__(self, text, deadline, created=datetime.datetime.now()):
+        self.text = text
+        self.deadline = datetime.timedelta(deadline)
+        self.created = datetime.datetime.now()
+
+    def is_active(self):
+        time_now = datetime.datetime.now()
+        time_deadline = self.created + self.deadline
+        return False if time_now >= time_deadline else True
+
+
+# пока почему-то не работает
+class DeadlineError(Exception):
+    """
+    Initiate Exception class
+    """
+
+    def __init__(self, method_is_active):
+        self.is_active = method_is_active
+
+    def __str__(self):
+        try:
+            assert self.is_active is True
+        except AssertionError:
+            raise DeadlineError
+        else:
+            pass
```

Comment:
Исключения довольно часто не требуют никакого кода, достаточно просто иметь отдельный класс исключения что бы потом его отлавливать и всё. Так что можно пропустить тело класса вообще
```suggestion
    ...
```